### DISCUSSION TITLE
Basic question - I've solved a scalar Laplacian and I want to output the gradient

### DISCUSSION THREAD

**Discussion by ethrelfall** [firedrakeproject/firedrake#3590] [Firedrake support] [answered]
Dear community,

Apologies in advance for what's probably a silly question.

I've solved a Laplace eq for a potential phi (let's say it's on a function space of CG 1 elements); I can output phi fine, but how do I output grad(phi)?  The interface doesn't allow me to set up a Function (on whatever space) and interpolate grad(phi), saying "cannot interpolate an expression with an argument into a Function".

Also I'd like to output grad phi to the accuracy of the phi solve (i.e. not enforce continuity in the gradient that isn't there in the true gradient of phi).

Thanks!

(Search for matching discussions done - nada.)

**Comment by colinjcotter** [comment]
To do this exactly, find a finite element space that contains the gradient and interpolate to it. For the example of CG1, the gradient is in (rotated) RT0 but you might as well interpolate to DG1 which contains RT0 as a subspace.

**Comment by ethrelfall** [comment]
Thanks so much, I try

VV = VectorFunctionSpace(mesh,"DG",0)
stupid = Function(VV)
stupid.interpolate(grad(phi))

but I get that error can't interpolate an expression with an argument into a Function ... how do I do this pls?

(aside I guess rotated RT0 is RTE0 ... i.e. tangentially-continuous ...)

**Reply by colinjcotter** [reply]
huh! project should work here anyway, and is cheap for DG (you need DG1 though)

**Reply by pbrubeck** [reply]
Although the space you are looking for is `N1curl`, we can only output vector CG or DG functions to paraview. For the default Lagrange basis functions you can create the space as `VectorFunctionSpace(mesh, "DG", 1)` . You may also use Legendre polynomials with `VectorFunctionSpace(mesh, "DG", 1, variant="integral")` for which projection and interpolation are equivalent. Projection is a bit more expensive because it involves solving a system with a mass matrix.

**Reply by Alpha-Legion** [reply]
Now paraview seems support other types of finite elements. I've tried for RTE and it works well.

**Reply by pbrubeck** [reply]
Although paraview recently started supporting partially continuous vector fields, Firedrake does not fully exploit this feature. For non-Lagrange elements like N1curl or RT, `VTKFile.write` will `interpolate` the function into a DG element under the hood. So yes, paraview just "works" on any finite element, but it does not respect partial continuity.

**Reply by Alpha-Legion** [reply]
Oh, it was `VTKFile.write`'s good job! I just know that from you :) Thanks!

**Comment by rckirby** [comment, marked as answer]
Any chance a TrialFunction instead of a Function is being interpolated in the code?On May 24, 2024, at 4:51 PM, Colin J Cotter ***@***.***> wrote:﻿
huh! project should work here anyway, and is cheap for DG (you need DG1 though)

**Reply by ethrelfall** [reply]
Every chance - that was the problem ... sorry and thanks both!
